doc/man3/ASN1_item_sign.pod

```diff
@@ -0,0 +1,133 @@
+=pod
+
+=head1 NAME
+
+ASN1_item_sign, ASN1_item_sign_ex, ASN1_item_sign_ctx,
+ASN1_item_verify, ASN1_item_verify_ex, ASN1_item_verify_ctx -
+ASN1 sign and verify
+
+=head1 SYNOPSIS
+
+ #include <openssl/x509.h>
+
+ int ASN1_item_sign_ex(const ASN1_ITEM *it, X509_ALGOR *algor1,
+                       X509_ALGOR *algor2, ASN1_BIT_STRING *signature,
+                       const void *data, const ASN1_OCTET_STRING *id,
+                       EVP_PKEY *pkey, const EVP_MD *md, OSSL_LIB_CTX *libctx,
+                       const char *propq);
+
+ int ASN1_item_sign(const ASN1_ITEM *it, X509_ALGOR *algor1, X509_ALGOR *algor2,
+                    ASN1_BIT_STRING *signature, const void *data,
+                    EVP_PKEY *pkey, const EVP_MD *md);
+
+ int ASN1_item_sign_ctx(const ASN1_ITEM *it, X509_ALGOR *algor1,
+                        X509_ALGOR *algor2, ASN1_BIT_STRING *signature,
+                        const void *data, EVP_MD_CTX *ctx);
+
+ int ASN1_item_verify_ex(const ASN1_ITEM *it, const X509_ALGOR *alg,
+                         const ASN1_BIT_STRING *signature, const void *data,
+                         const ASN1_OCTET_STRING *id, EVP_PKEY *pkey,
+                         OSSL_LIB_CTX *libctx, const char *propq);
+
+ int ASN1_item_verify(const ASN1_ITEM *it, const X509_ALGOR *alg,
+                      const ASN1_BIT_STRING *signature, const void *data,
+                      EVP_PKEY *pkey);
+
+ int ASN1_item_verify_ctx(const ASN1_ITEM *it, const X509_ALGOR *alg,
+                          const ASN1_BIT_STRING *signature, const void *data,
+                          EVP_MD_CTX *ctx);
+
+=head1 DESCRIPTION
+
+ASN1_item_sign_ex() is used to sign arbitrary ASN1 data using a data object
+I<data>, the ASN.1 structure I<it>, algorithm identifiers I<algor1> and I<algor2>,
```

Comment:
What do algor1 and algor2 actually *do*?

Reply:
That has never been well explained, and I've seen use cases where they are used very differently.  Even looking through git history, it remains a mystery, at least as far as I was able to dig last time I tried to figure this out.

In X.509 terms, it looks like `algor1` is meant the get the algorithm from `signature` from `TBScertificate`, and `algor2` is meant to get the algorithm from `signatureAlgorithm` in `Certificate`.
(I still have a hard time understanding the difference between those two algos, so can't help further)

However, with other uses, those two get other algorithms.  As I recall, PKCS7 and CMS use these, with their own assumptions.

In the end, this depends hugely on the use, and the called of `ASN1_item_sign_ex()` *must* know the backend that does the grunt of the work intimately, and it all comes down to which ctrl commands are used.

Reply:
They are output parameters, not input. The to be signed data, and the struct with the signature can both have copy of the X509_ALGOR. The ASN1_item_sign() function will write the correct X509_ALGOR to the structs based on the algorithms and parameters that have been set up. If one of the algors points to the X509_ALGOR of the TBS data, that X509_ALGOR will first be written before the signature is generated.